pkg/model/diamond.go

```diff
@@ -0,0 +1,265 @@
+	ForbidConflicts ConflictMode = "forbids-conflicts"
+)
+
+// DiamondState models the running status of an ungoing diamond workflow
+type DiamondState string
+
+const (
+	// DiamondInitialized is the state of an initialized diamond
+	DiamondInitialized DiamondState = "initialized"
+
+	// DiamondCommitting indicates that a commit is going on for this diamond
+	DiamondCommitting DiamondState = "committing"
+
+	// DiamondCancelling indicates that a cancel is going on for this diamond
+	DiamondCancelling DiamondState = "cancelling"
+
+	// DiamondRunning indicates that the diamond has started some splits
+	DiamondRunning DiamondState = "running"
+
+	// DiamondDone indicates the diamond has completed with a commit
+	DiamondDone DiamondState = "done"
+
+	// DiamondCancelled indicates the diamond has completed with a cancel
+	DiamondCancelled DiamondState = "cancelled"
+)
+
+const (
+	diamondDescriptorFile = "diamond.yaml"
+	splitDescriptorFile   = "split.yaml"
+)
+
+// IsValid checks the value of a diamond state
+func (s DiamondState) IsValid() bool {
+	switch s {
+	case DiamondInitialized, DiamondRunning, DiamondCommitting, DiamondDone, DiamondCancelling, DiamondCancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s DiamondState) String() string {
+	return string(s)
+}
+
+// SplitState models the running status of an ungoing diamond split
+type SplitState string
+
+const (
+	// SplitDone is the state of a completed split
+	SplitDone SplitState = "done"
+
+	// SplitFailed is the state of a failed split
+	SplitFailed SplitState = "failed"
+
+	// SplitRunning is the state of running split
+	SplitRunning SplitState = "running"
+)
+
+// IsValid checks the value of a diamond state
+func (s SplitState) IsValid() bool {
+	switch s {
+	case SplitDone, SplitFailed, SplitRunning:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s SplitState) String() string {
+	return string(s)
+}
+
+// DiamondDescriptor models a diamond's metadata
+type DiamondDescriptor struct {
+	DiamondID       string            `json:"diamondID" yaml:"diamondID"`
+	StartTime       time.Time         `json:"startTime" yaml:"startTime"` // when the diamond was initialized
+	EndTime         time.Time         `json:"endTime,omitempty" yaml:"endTime,omitempty"`
+	StateChangeTime time.Time         `json:"stateChangeTime" yaml:"stateChangeTime"` // last state change
+	State           DiamondState      `json:"state" yaml:"state"`
+	Mode            ConflictMode      `json:"mode" yaml:"mode"`
+	HasConflicts    bool              `json:"hasConflicts,omitempty" yaml:"hasConflicts,omitempty"`
+	HasCheckpoints  bool              `json:"hasCheckpoints,omitempty" yaml:"hasCheckpoints,omitempty"`
+	Tag             string            `json:"tag,omitempty" yaml:"tag,omitempty"`
+	BundleID        string            `json:"bundleID,omitempty" yaml:"bundleID,omitempty"`
+	Splits          []SplitDescriptor `json:"splits,omitempty" yaml:"splits,omitempty"`
```

Comment:
aside from the general motif of separating out data (struct fields) according to mutability expectations/guarantees , the `Splits` field is the most surprising part of the `DiamondDescriptor` to me, because we'll also be writing

```
/diamonds/{repo}/{diamond-id}/splits/{split-id}/split.yaml
/diamonds/{repo}/{diamond-id}/splits/{split-id}/bundle-files-{index}.yaml
```

on `split add`.

reading ahead, i can see this field is set only upon `diamond commit`, not `split add` (i'm pretty sure?), and going further, i'm pretty sure that it's only accessed during that operation as well such that it needn't necessarily be part of this struct?

this is all happening in vmetadata in any case, so the possibility for inconsistency by writing the same data in two places isn't particularly disturbing, particularly if we get the GC-like pod going sooner rather that later, since that program can also do things like check consistency / validate vmetadata.